ulauncher/modes/extensions/extension_environment.py

```diff
@@ -0,0 +1,58 @@
+"""
+!EXPERIMENTAL: this module intends to provide an implementation of automatic
+!installation and virtual environment management/setup automation of ulauncher
+!extensions installed using one way or another.
+"""
+
+from pathlib import Path
+from os.path import exists
+
+
+class ExtensionEnvironment:
+    def __init__(self, ext_dir: Path, venv_map: dict[str, set]) -> None:
+        self.ext_dir: Path = ext_dir
+        self.venv_map: dict[str, set[str]] = {
+            "test": set(
+                    [
+                        "hello==1.0"
+                    ]
+                )
+        }
+
+    def check_requirements(self) -> bool:
+        """check_requirements
+
+        Args:
+            ext_dir (Path): Path of the extension dir downloaded/cloned.
+
+        Returns:
+            bool: whether a requirements.txt file exists for the extension.
+        """
+
+        if exists(self.ext_dir / "requirements.txt"):
+            return True # it should be requirements.txt only, not requirement.txt
+        return False
+
+    def prep_extension(self):
+        req_path: Path = self.ext_dir / "requirements.txt"
+
+        if not self.check_requirements():
+            self.create_sole_venv()
```

Comment:
If there is no requirements.txt then there probably shouldn't be a venv.

Reply:
well, i think a general/isolatd venv for an extension with no requirements.txt will be better than none at all, so even if it messes up, it can be contained

Reply:
Note that a Python virtual environment has no meaningful isolation beyond the process isolation that is already present. Despite the name, it doesn't imply anything akin to a java virtual machine or similar. It really is just a directory of modules, and the PYTHONPATH is modified to point to those modules in preference to the system modules. Generally the Python executable in the `bin` path of the venv is just a wrapper script which sets up the path information and then exec the actual system Python executable.